sqlx-core/src/pool/inner.rs

```diff
@@ -300,18 +301,31 @@ impl<DB: Database> PoolInner<DB> {
 
         let mut backoff = Duration::from_millis(10);
         let max_backoff = deadline_as_timeout::<DB>(deadline)? / 5;
+        let mut num_attempts: u32 = 0;
 
         loop {
             let timeout = deadline_as_timeout::<DB>(deadline)?;
+            num_attempts += 1;
 
             // clone the connect options arc so it can be used without holding the RwLockReadGuard
             // across an async await point
-            let connect_options = self
+            let connect_options_arc = self
                 .connect_options
                 .read()
                 .expect("write-lock holder panicked")
                 .clone();
 
+            let connect_options = if let Some(callback) = &self.options.before_connect {
+                callback(connect_options_arc.as_ref(), num_attempts)
```

Comment:
Hmm, this should probably be included in the timeout if it's intended for anything non-trivial.

Reply:
This would mean that you can't separate the timeout for the database connection from the timeout from this callback. My thinking here was that the external call is under the control of the caller, and so they could implement appropriate timeouts externally.

I see your point of view too. I'll implement this change, but shout if the above argument sways you.

Reply:
After this change, the behavior is slightly different; the code falls into the same error handling for `before_connect` as the rest of the connect code; this means a `ConnectionRefused` error causes a retry. This seems okay, for the same reasons that sqlx retries in these cases.
Second behavior change: getting the connection options from the shared mutex is included in the timeout calculation. This is very unlikely to significantly affect the outcome; and if it did, the argument for including it is the same as the argument for including the time for the callback to run.

Reply:
My concern for _not_ including it in the timeout is the user forgetting to implement their own timeout and a call stalling forever.

> Second behavior change: getting the connection options from the shared mutex is included in the timeout calculation. 

That's a non-issue; the critical section is very short.